### PR TITLE
wabt rethrow exception in finish

### PR DESCRIPTION
**Pull request by poemm**
Exceptions in Hera would not allow Wabt invocation to free anything, resulting in a memory leak. This pull request catches and records Hera exceptions, uses https://github.com/WebAssembly/wabt/pull/886 to allow Wabt to clean up, then rethrows the Hera exception.

Implemented for the `finish` eei method, still needs to be implemented for the rest.

### CODE REVIEW

File: src/wabt.cpp [hunk 1]
```diff
@@ -57,12 +57,15 @@ class WabtEthereumInterface : public EthereumInterface {
     m_wasmMemory = _wasmMemory;
   }
 
+  std::exception_ptr my_exception_ptr;
```
Comment (axic) [inline src/wabt.cpp]: Can you please not go with `my_exception_ptr`, rather something like `last_exception_ptr` ?

File: src/wabt.cpp [hunk 2]
```diff
@@ -104,7 +107,15 @@ ExecutionResult WabtEngine::execute(
     ) {
       int64_t gas = static_cast<int64_t>(args[0].value.i64);
       // FIXME: handle host trap here
-      interface.eeiUseGas(gas);
+      try {
+        interface.eeiUseGas(gas);
+      } catch (HeraException const& e) {
```
Comment (axic) [inline src/wabt.cpp]: There's no need to name these arguments.

File: src/wabt.cpp (hunk 1 above)
Comment (axic) [inline src/wabt.cpp]: No need to prefix with `std::`

File: src/wabt.cpp [hunk 3]
```diff
@@ -260,14 +333,22 @@ ExecutionResult WabtEngine::execute(
   // FIXME: really bad design
   interface.setWasmMemory(env.GetMemory(0));
 
+  // for rethrowing exception
+  interface.my_exception_ptr = nullptr;
```
Comment (axic) [inline src/wabt.cpp]: I don't think this needs to be null'd here.

**Comment by axic** [comment]
Did anybody check if wabt isn't already rethrowing non-wabt exceptions?

File: src/wabt.cpp (hunk 1 above)
Comment (poemm) [reply src/wabt.cpp]: Ok.

File: src/wabt.cpp (hunk 2 above)
Comment (poemm) [reply src/wabt.cpp]: Ok.

File: src/wabt.cpp (hunk 1 above)
Comment (poemm) [reply src/wabt.cpp]: Ok.

File: src/wabt.cpp (hunk 3 above)
Comment (poemm) [reply src/wabt.cpp]: Ok.

**Comment by poemm** [comment]
> Did anybody check if wabt isn't already rethrowing non-wabt exceptions?

They don't.

**Comment by poemm** [comment]
@chfast Please confirm that there are no detected memory leaks.

**Comment by poemm** [comment]
(Responding to https://github.com/ewasm/hera/pull/473#issuecomment-451942837 .)

> Let's re-review it. Maybe there is something valuable there.

@chfast I re-reviewed this and don't see anything else interesting here. I think that this PR was useful because it showed that there is a way to solve memory leaks, which give you a hint for your better solution.

**Comment by chfast** [comment]
>  I re-reviewed this and don't see anything else interesting here. I think that this PR was useful because it showed that there is a way to solve memory leaks, which give you a hint for your better solution.

Yes, I'm for closing this one.

**Pull request closed without merging**